// control-id/objects/groups/Groups.js
const axios = require("axios");
const http = require("http");
const Login = require("../../login/Login");

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class Groups {
  /**
   * @argument {string} ip host em que se encontra o painel do controlID
   */
  constructor(ip,login) {
    this.__ip = ip; // HOST
    this._session = login;
  }

  /**
   * @param {number} nome nome do grupo que deseja criar.
   * @returns
   */
  async createGroup(nome) {
    try {
      return new Promise(async (resolve, reject) => {
        try {
          const response = await instance.post(
            "http://" + this.__ip + "/create_objects.fcgi?session=" + this._session,
            {
              "object": "groups",
              "values": [
                {
                  "name": nome,
                },
              ],
            },
            {
              headers: {
                "Content-Type": "application/json",
              },
            },
          );

          resolve(response.data);
        } catch (error) {
          reject(new Error(error.response.data.error));
        }
      });
    } catch (error) {
      throw new Error(error.message);
    }
  }

  /**
   * @description acessar as regras salvas
   * @returns
   */
  async loadGroup() {
    try {
      return new Promise(async (resolve, reject) => {
        try {
          const response = await instance.post(
            "http://" + this.__ip + "/load_objects.fcgi?session=" + this._session,
            {
              "object": "groups",
            },
            {
              headers: {
                "Content-Type": "application/json",
              },
            },
          );

          resolve(response.data); 
        } catch (error) {
          reject(new Error(error.response.data.error));
        }
      });
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  /**
   * @param {number} group_id id do grupo que deseja criar
   * @returns
   */
  async deleteGroup(group_id) {
    try {
      return new Promise(async (resolve, reject) => {
        try {
          const response = await instance.post(
            "http://" + this.__ip + "/destroy_objects.fcgi?session=" + this._session,
            {
              "object": "groups",
              "where": {
                "groups": {
                  "id": group_id,
                },
              },
            },
            {
              headers: {
                "Content-Type": "application/json",
              },
            },
          );

          resolve(response.data);
        } catch (error) {
          reject(new Error(error.response.data.error));
        }
      });
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }
}

(
  async function groupLogin(){
    try {
  
      const Grupo = new Groups('192.168.88.10');
  
      console.log( await Grupo.loadGroup())
      
    } catch (error) {
      console.error(error)
    }
  }
)()
module.exports = Groups;
