// control-id/users/User.js
const axios = require("axios");
const http = require("http");
const Login = require("../login/Login");

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class Users {
  /**
   * @param {string} ip The IP address of the host
   * @param {string} login The login Session token 
   */
  constructor(ip, login) {
    this.__ip = ip;
    this._session = login;
  }

  async addUser(nome, matricula, dataInicial, dataFinal, id) {
    // console.log(`Nome: ${nome}, matricula: ${matricula}, dataInit: ${dataInicial}, dataFim: ${dataFinal}, id: ${id}`)
    try {
      const dataTime = new Date(dataInicial);
      const endTme = new Date(dataFinal);
      const response = await instance.post(
        `http://${this.__ip}/create_objects.fcgi?session=${this._session}`,
        {
          object: "users",
          values: [
            {
              id: id,
              name: nome,
              registration: `${matricula}`,
              begin_time: parseInt(Math.floor(dataTime.getTime() / 1000)),
              end_time: parseInt(Math.floor(endTme.getTime() / 1000)),
            },
          ],
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
        
      return response.data;
    } catch (e) {
      if (e.response && e.response.data) {
        console.log(e.response.data);
        throw new Error(e.response.data.error);
      } else {
        console.log(e);
        throw new Error("Ocorreu um erro desconhecido.");
      }
    }
  }

  async addGroup(id) {
    try {
      const response = await instance.post(
        `http://${this.__ip}/create_objects.fcgi?session=${this._session}`,
        {
          object: "user_groups",
          fields:[ "user_id", "group_id" ],
          values: [
            {
              user_id: id,
              group_id: 1
            },
          ],
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
        
      return response.data;
    } catch (e) {
      if (e.response && e.response.data) {
        console.log(e.response.data);
        throw new Error(e.response.data.error);
      } else {
        console.log(e);
        throw new Error("Ocorreu um erro desconhecido.");
      }
    }
  }

  /**
   * 
   * @returns {{
   *  users: [
   *  {
        id: int,
        registration: string,
        name: string,
        password: string,
        panic_password: string,
        salt: string,
        panic_salt: string,
        expires: int,
        user_type_id: int,
        begin_time: int,
        end_time: int,
        image_timestamp: int,
        last_access: int
      }]
   * }} An object containing all the users.
   */
  async getUsers() {
    try {
      const response = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        { object: "users" }
      );
      return response.data 
    } catch (error) {
      return null
      // console.error(`Erro getUsers no IP ${this.__ip}:`, error.message);
      // return { success: false, error: error.response?.data?.error || error.message };
    }
  }
  

  /**
   * 
   * @param {int} user_id The id of the user
   * @returns {{
      id: int,
      registration: string,
      name: string,
      password: string,
      panic_password: string,
      salt: string,
      panic_salt: string,
      expires: int,
      user_type_id: int,
      begin_time: int,
      end_time: int,
      image_timestamp: int,
      last_access: int
    }} The user object
   */
  async getUser(user_id) {
    try {
      const response = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        {
          object: "users",
          where: {
            users: {
              id: parseInt(user_id),
            },
          },
        }
      );
      return response.data.users[0];
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  /**
   * 
   * @param {string} registration The registration of the user
   * @returns {{
      id: int,
      registration: string,
      name: string,
      password: string,
      panic_password: string,
      salt: string,
      panic_salt: string,
      expires: int,
      user_type_id: int,
      begin_time: int,
      end_time: int,
      image_timestamp: int,
      last_access: int
    }} The user object
   */
  async getUserByRegistration(registration) {
    try {
      const response = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        {
          object: "users",
          where: {
            users: {
              registration: registration
            },
          },
        }
      );
      return response.data.users[0];
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async deleteUsers() {
    try {
      const response = await instance.post(
        `http://${this.__ip}/destroy_objects.fcgi?session=${this._session}`,
        {
          object: "users",
        }
      );
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  

  async deleteUser(id_user) {
    try {
      const response = await instance.post(
        `http://${this.__ip}/destroy_objects.fcgi?session=${this._session}`,
        {
          object: "users",
          where: {
            users: {
              id: parseInt(id_user),
            },
          },
        }
      );
      console.log(response.data)
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async updateUser(id_user, matricula, nome, dataInicial, dataFinal) {
    
    try {
      const dataTime = new Date(dataInicial);
      const endTme = new Date(dataFinal);

      const response = await instance.post(
        `http://${this.__ip}/modify_objects.fcgi?session=${this._session}`,
        {
          object: "users",
          values: {
            begin_time: dataTime,
            end_time: endTme,
          },
          where: {
            users: {
              id: id_user,
            },
          },
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async updateCodeUser(registration, id_user) {
    
    try {
      const response = await instance.post(
        `http://${this.__ip}/modify_objects.fcgi?session=${this._session}`,
        {
          object: "users",
          values: {
            id: parseInt(id_user),
          },
          where: {
            users: {
              registration: registration,
            },
          },
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async getUsersRules() {
    
    try {
      const response = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        {
          object: "user_groups",
        }
      );
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async createUserGroup(user_Id, group_id) {
    
    try {
      const response = await instance.post(
        `http://${this.__ip}/create_objects.fcgi?session=${this._session}`,
        {
          object: "user_groups",
          fields: ["user_id", "group_id"],
          values: [
            {
              user_id: user_Id,
              group_id: group_id,
            },
          ],
        }
      );
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async blockUser(user_id, group_id) {
    
    try {
      const host = this.__ip;
      const session = this._session;

      const response1 = await instance.post(
        `http://${host}/destroy_objects.fcgi?session=${session}`,
        {
          object: "user_groups",
          where: {
            user_groups: {
              user_id: {
                "=": parseInt(user_id),
              },
            },
          },
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      const response2 = await instance.post(
        `http://${this.__ip}/create_objects.fcgi?session=${this._session}`,
        {
          object: "user_groups",
          fields: ["user_id", "group_id"],
          values: [
            {
              user_id: parseInt(user_id),
              group_id: parseInt(group_id),
            },
          ],
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      const resposta1 = response1.data;
      const resposta2 = response2.data;
      return [
        {
          user_groups: resposta1,
          user_access_rules: resposta2,
        },
      ];
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async getUserAccessGroup() {
    
    try {
      const response = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        {
          object: "access_rules",
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async unblockUser(id_user, group_id, access_rule_id) {
    
    try {
      await instance.post(
        `http://${this.__ip}/destroy_objects.fcgi?session=${this._session}`,
        {
          object: "user_groups",
          where: {
            user_groups: {
              user_id: {
                "=": parseInt(id_user),
              },
            },
          },
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      const response1 = await instance.post(
        `http://${this.__ip}/create_objects.fcgi?session=${this._session}`,
        {
          object: "user_groups",
          fields: ["user_id", "group_id"],
          values: [
            {
              user_id: parseInt(id_user),
              group_id: parseInt(group_id),
            },
          ],
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      const response2 = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        {
          object: "user_access_rules",
          values: [
            {
              user_id: parseInt(id_user),
              access_rule_id: 1,
            },
          ],
        },
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      return [response1.data, response2.data];
    } catch (error) {
      console.log(error)
      throw new Error(error.response.data.error);
    }
  }
}

const req = new Users('192.168.88.11');

async function getUserByRegister(){
  try {
    const result = await req.getUser("694")

    console.log(result)
  } catch (error) {
    console.log(error)
  }
}
// getUserByRegister()
module.exports = Users;
