// control-id/system/System.js
const axios = require("axios");
const http = require("http");
const Login = require("../login/Login");

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class System {
  /**
   * @param {string} ip The IP address of the host
   * @param {string} login The login Session token 
   */
  constructor(ip, login) {
    this.__ip = ip; // HOST
    this._session = login;
  }

  async getSystemInfo() {
    const endpoint = `http://${this.__ip}/system_information.fcgi?session=${this._session}`;
    const headers = { "Content-Type": "application/json" };

    try {
      const response = await instance.post(endpoint, { headers });
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  /**
   * @description Pegar todos os logs no controle de acesso.
   * @returns
   */
  async getSystemLogs() {
    const endpoint = `http://${this.__ip}/get_ac_log.fcgi?session=${this._session}`;
    const headers = { "Content-Type": "application/json" };

    try {
      const response = await instance.post(endpoint, { headers });
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async getAccessLogs() {
    const endpoint = `http://${this.__ip}/load_objects.fcgi?session=${this._session}`;
    const headers = { "Content-Type": "application/json" };
    const dataBody = {
      object: "access_logs",
      name: "Logs de Acesso",
      order: ["id"],
      fields: {
        id: { label: "Código", value: "", type: "int", not_null: false, PK: true, show: false },
        time: { label: "Tempo", type: "time", not_null: false },
        user_id: { type: "int", show: false },
        portal_id: { type: "int", show: false },
        users: { label: "Usuário", value: "", type: "fields", isField: false, not_null: false, showTable: false, fieldType: "single", fieldId: "user_id" },
        portals: { label: "Portal", value: "", type: "fields", isField: false, not_null: false, showTable: false, fieldType: "single", fieldId: "portal_id" },
        areas: { label: "Área", object: "areas", type: "fields", isField: false, not_null: false, showTable: false, fieldType: "single" },
        event: { label: "Evento", type: "int", not_null: false, show: false },
      },
    };

    try {
      const response = await instance.post(endpoint, dataBody, { headers });
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }
}

module.exports = System;
