// control-id/objects/card/Card.js
const axios = require("axios");
const http = require("http");
const Login = require("../../login/Login");

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class Card {
  /**
   * @param {string} ip The IP address of the host
   * @param {string} login The login Session token 
   */
  constructor(ip, login) {
    this.__ip = ip;
    this._session = login;
  }

  async createCard(user_id, card_decimal) {
    const card_decimalComZeros = adicionarZerosEsquerda(card_decimal, 8);
    const { area, codigo } = separarCodigoEArea(card_decimalComZeros);
    const card_int_value = area * Math.pow(2, 32) + codigo;
    const endpoint = `http://${this.__ip}/create_objects.fcgi?session=${this._session}`;
    const requestData = {
      object: "cards",
      values: [{ value: card_int_value, user_id: user_id }],
    };

    try {
      const response = await instance.post(endpoint, requestData, { headers: { "Content-Type": "application/json" } });
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async createCardConverting(user_id, card_decimal) {
    const { upper16Bits, lower16Bits } = convertRFIDFormats(card_decimal);
    const card_int_value = upper16Bits * Math.pow(2, 32) + lower16Bits;

    try {

      const response = await instance.post(
        `http://${this.__ip}/create_objects.fcgi?session=${this._session}`,
        { object: "cards", values: [{ value: card_int_value, user_id: user_id }] },
        { headers: { "Content-Type": "application/json" } }
      );

      return response.data;
    } catch (error) {
      console.log(error.response.data)
      throw new Error(error.response.data.error);
    }
  }

  deleteCard(id_user) {
    return new Promise(async (resolve, reject) => {
      try {
        const response = await instance.post(
          `http://${this.__ip}/destroy_objects.fcgi?session=${this._session}`,
          { object: "cards", where: { cards: { user_id: parseInt(id_user) } } },
          { headers: { "Content-Type": "application/json" } }
        );

        resolve(response.data);
      } catch (error) {
        reject(error.response.data.error);
      }
    });
  }

  async updateCard(id_user, card_value) {
    try {
      await this.DeletCard(id_user);
      await this.CriarCardConvert(id_user, Number(card_value));
    } catch (error) {
      console.log(error)
      throw new Error(error.message);
    }
  }

  async LoadCard(id_user) {
    try {
      const response = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        { object: "cards", where: { cards: { user_id: parseInt(id_user) } } },
        { headers: { "Content-Type": "application/json" } }
      );

      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }

  async getCards() {
    try {
      const response = await instance.post(
        `http://${this.__ip}/load_objects.fcgi?session=${this._session}`,
        { object: "cards" },
        { headers: { "Content-Type": "application/json" } }
      );

      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }
}

module.exports = Card;


async function listCar() {
  try {

    const card = new Card('192.168.88.10');

    console.log(await card.LoadCard(2907))
    
  } catch (error) {
    console.log(error)
  }
  
}

listCar()

// Funções auxiliares
function adicionarZerosEsquerda(numero, tamanho) {
  const numeroString = numero.toString();
  const zerosFaltantes = tamanho - numeroString.length;
  return zerosFaltantes > 0 ? "0".repeat(zerosFaltantes) + numeroString : numeroString;
}

function separarCodigoEArea(card_decimal) {
  const codigo = card_decimal % 100000;
  const area = Math.floor(card_decimal / 100000);
  return { area, codigo };
}

function convertRFIDFormats(rfidNumber) {
  const decimalValue = parseInt(rfidNumber, 10);
  const hexadecimalValue = decimalValue.toString(16).toUpperCase();
  const binaryValue = decimalValue.toString(2);
  const upper16Bits = (decimalValue >> 16) & 0xffff;
  const lower16Bits = decimalValue & 0xffff;
  return {
    decimal: decimalValue,
    hexadecimal: hexadecimalValue,
    binary: binaryValue,
    upper16Bits: upper16Bits,
    lower16Bits: lower16Bits,
  };
}
