// control-id/objects/export_log/Data.js
const axios = require("axios");
const http = require("http");
const Login = require("../../login/Login")
const moment = require('moment')

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class ExportLogData {
  /**
   * @param {string} ip The IP address of the host
   * @param {string} login The login Session token 
   */
  constructor(ip, login) {
    this.__ip = ip; // HOST
    this._session = login;
  }

  async exportDataFromDateRange(startDate, endDate) {
    try {
      const endpoint = `http://${this.__ip}/load_objects.fcgi?session=${this._session}`;
      const requestData = { 
        join: "LEFT", 
        object: "access_logs", 
        fields: [], 
        where: [ 
          {
            field: "time",
            operator: ">=",
            value: moment(startDate, "YYYY-MM-DD").startOf("day").unix()
          },
          {
            field: "time",
            operator: "<=",
            value: moment(endDate, "YYYY-MM-DD").endOf("day").unix()
          }
        ], 
        order: ["id"], 
        offset: 0 
      };
      

      const headers = {
        "Content-Type": "application/json",
      }; 

      const response = await instance.post(endpoint, requestData, { headers });
      return response.data;

    } catch (error) {
      console.error(error);
      // throw new Error(error.response ? error.response.data.error : error.message);
      return null
    }
  }


  async loadObjects() {
    const endpoint = `http://${this.__ip}/load_objects.fcgi?session=${this._session}`;
    const requestData = {
      'object': 'report_filters',
      //   'where' : {
      //    'report_filters' :{
      //      'report_id' : parseInt(queryString.report)
      //    }
      //  }
    }
    const headers = {
      "Content-Type": "application/json",
    };

    try {
      const response = await instance.post(endpoint, requestData, { headers });

      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }
}


// const exporte = new Export('192.168.88.10');

// async function teste(){
//   try {

//     const response = await exporte.loading_object();

//     console.log(response)
//   } catch (error) {

//     console.error(error)
//   }
// }
// teste()

module.exports = ExportLogData;
