// control-id/objects/export_log/ExportLog.js
const axios = require("axios");
const http = require("http");
const Login = require("../../login/Login");

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class ExportLog {
  /**
   * @param {string} ip The IP address of the host
   * @param {string} login The login Session token 
   */
  constructor(ip,login) {
    this.__ip = ip; // HOST
    this._session = login;
  }

  async exportData() {

    const endpoint = `http://${this.__ip}/export_objects.fcgi?session=${this._session}`;
    const requestData = {
      objects: [
        {
          object: "access_logs",
          // name: "Logs de Acesso",
          // order: ["id"],
          // fields: {
          //   id: { label: "Código", value: "", type: "int", not_null: false, PK: true, show: false },
          //   time: { label: "Tempo", type: "time", not_null: false },
          //   user_id: { type: "int", show: true },
          //   users: { label: "Usuário", value: "", type: "fields", isField: false, not_null: false, showTable: false, fieldType: "single", fieldId: "user_id" },
          //   portals: { label: "Portal", value: "", type: "fields", isField: false, not_null: false, showTable: false, fieldType: "single", fieldId: "portal_id",
          //   },
          //   areas: { label: "Área", object: "areas", type: "fields", isField: false, not_null: false, showTable: false, fieldType: "single" },
          //   event: { label: "Evento", type: "int", not_null: false, show: false },
          // },
        },
      ],
    };

    const headers = {
      "Content-Type": "application/json",
    };

    try {
      const response = await instance.post(endpoint, requestData, { headers });
      return response.data;
    } catch (error) {
      throw new Error(error.response.data.error);
    }
  }
}

const exporte = new ExportLog("192.168.88.10");

module.exports = ExportLog;
