// control-id/facial_recognition/FacialRecognition.js
const axios = require("axios");
const http = require("http");
const moment = require('moment');

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class FacialRecognition {
    /** @private */ __ip;
    /** @private */ _session;

    
    constructor(ip, login){
        this.__ip = ip;
        this._session = login;
    }

    /**
     * 
     * Uploads an image to serve as the face recognition parameter
     * 
     * Usage:
     * ```js
     *  let login = await new Login("10.10.10.10", "<user>", "<password>").login();
     *  let fr = new FactialRecognition("10.10.10.10",login);
     *  await fr.registerUserImage(<match>,<user_id>,<stream>);
     * ```
     * 
     * @param {boolean} match Option to match against existing images, failing to register if the face is found.
     * @param {int} user_id  The user that to map the face to.
     * @param {Buffer} stream The image as a buffer.
     * 
     */
    async registerUserImage(match, user_id, stream){
        try {
            const endpoint = `http://${this.__ip}/user_set_image.fcgi?user_id=${user_id}&match=${match ? "1" : "0"}&timestamp=${moment().unix()}&session=${this._session}`;
            const requestData = stream
            const headers = {
                "Content-Type": "application/octet-stream",
            };

            const res = await instance.post(endpoint, requestData, { headers });
            if(res.data.errors){
                throw new Error("Não foi possivel registrar Rosto: " + res.data.errors[0].message + ", error code:" + res.data.errors[0].code);
            }
            return res.data;
        } catch (e) {
            throw new Error(e.message);
        }
    }

    /**
     * 
     * Gets the image of the User with the id provided.
     * 
     * Usage:
     * ```js
     *  let login = await new Login("10.10.10.10", "<user>", "<password>").login();
     *  let fr = new FactialRecognition("10.10.10.10",login);
     *  let data = await fr.getUserImage(<user_id>);
     * 
     *  // Handling
     * ```
     * 
     * @returns {{timestamp:int; image:string;}} An object contains the timestamp of when the image was registered and the image in Base64;
     */
    async getUserImage(user_id){
         try {
            const endpoint = `http://${this.__ip}/user_get_image.fcgi?user_id=${user_id}&get_timestamp=1&session=${this._session}`;
            
            const headers = {
                "Content-Type": "text/plain",
            };

            const res = await instance.get(endpoint, { headers });


            return res.data;
        } catch (e) {
            throw new Error(e.response.data.error);
        }
    }

/**
     * 
     * Gets the images of the Users with the ids provided.
     * 
     * 
     * @param {int[]} user_ids 
     * Usage:
     * ```js
     *  let login = await new Login("10.10.10.10", "<user>", "<password>").login();
     *  let fr = new FactialRecognition("10.10.10.10",login);
     *  let data = await fr.getUsersImageInBatch(<user_ids>);
     * 
     *  // Handling
     * ```
     * 
     * @returns {{timestamp:int; image:string;}} An object contains the timestamp of when the image was registered and the image in Base64;
     */
    async getUsersImageInBatch(user_ids){
         try {
            const endpoint = `http://${this.__ip}/user_get_image_list.fcgi?session=${this._session}`;
            
            const headers = {
                "Content-Type": "application/json",
            };

            const requestData = {
                "user_ids": user_ids
            };

            const res = await instance.post(endpoint, requestData, { headers });

            return res.data.user_images;
        } catch (e) {
            if(e.response){
                throw new Error(e.response.data.error);
            }else{
                throw new Error(e.message);
            }
        }
    }


    
    /**
     * 
     * Gets all the users that have an image registered
     * 
     * Usage:
     * ```js
     *  let login = await new Login("10.10.10.10", "<user>", "<password>").login();
     *  let fr = new FactialRecognition("10.10.10.10",login);
     *  let data = await fr.getRegisteredUsers();
     * 
     *  // Handling
     * ```
     * 
     * @returns {{image_info: [ {user_id:int; timestamp: int;}]}} An object that contains a list with all the users with a registered image;
     */
    async getRegisteredUsers(){
        try {
            const endpoint = `http://${this.__ip}/user_list_images.fcgi?get_timestamp=1&session=${this._session}`;
            const headers = {
                "Content-Type": "application/json",
            };

            const res = await instance.post(endpoint, { headers });


            return res.data;
        } catch (e) {
            throw new Error(e.response.data.error);
        }
    }

    /**
     * 
     * Tests a user's image
     * 
     * Usage:
     * ```js
     *  let login = await new Login("10.10.10.10", "<user>", "<password>").login();
     *  let fr = new FactialRecognition("10.10.10.10",login);
     *  await fr.testUserImage(<stream>);
     * ```
     * 
     * @param {Buffer} stream The image as a buffer.
     * 
     * @returns {{
            scores: {
                bounds_width: int,
                horizontal_center_offset: int,
                vertical_center_offset: int,
                center_pose_quality: int,
                sharpness_quality: int
            },
            success: boolean
        }} Returns the score of the match and if it successfully matched against any user
     */
    async testUserImage(stream) {
        try {
            const endpoint = `http://${this.__ip}/user_test_image.fcgi?get_timestamp=1&session=${this._session}`;
            const requestData = stream
            const headers = {
                "Content-Type": "application/octet-stream",
            };

            const res = await instance.post(endpoint, requestData, { headers });


            return res.data;
        } catch (e) {
            throw new Error(e.response.data.error);
        }
    }

    /**
     * 
     * Gets all the users that have an image registered
     * 
     * Usage:
     * ```js
     *  let login = await new Login("10.10.10.10", "<user>", "<password>").login();
     *  let fr = new FactialRecognition("10.10.10.10",login);
     *  let data = await fr.deleteUserImage(<user_id>);
     * 
     *  // Handling
     * ```
     */
    async deleteUserImage(user_id) {
        try {
            const endpoint = `http://${this.__ip}/user_destroy_image.fcgi?session=${this._session}`;
            const requestData = {
                "user_ids": [user_id]
            }
            const headers = {
                "Content-Type": "application/json",
            };

            const res = await instance.post(endpoint, requestData, { headers });


            return res.data;
        } catch (e) {
            throw new Error(e.response.data.error);
        }
    }

    /**
     * Takes a picture of the current frame in the camera of the reader.
     * 
     * @param {"rgb" | "ir"} [type="rgb"] The sensor used.
     * @returns {ArrayBuffer} The image.
     */
    async getFrameFromCamera(type = "rgb"){
        try {
            const endpoint = `http://${this.__ip}/save_screenshot.fcgi?session=${this._session}`;
            
            const requestData = {
                "frame_type": "camera",
                "camera": "rgb"
            }
            const headers = {
                "Content-Type": "application/json",
            };

            const res = await instance.post(endpoint, requestData, { headers, 
                responseType:"arraybuffer"
            });

            return res.data;
        } catch (e) {
            console.log(e)
            throw new Error(e.response.data.error);
        }
    }
}


module.exports = FacialRecognition;
