// control-id/config_monitor/configureMonitor.js
const axios = require("axios");
const http = require("http");
const Login = require("../login/login");


const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"]
  },
  insecureHTTPParser: true,
});

class Monitor {
    constructor(ip) {
        this.__ip = ip;
        this._session = null;
    }

    async configure(hostName, port){
         try {
            this._session = await new Login(this.__ip).login(); // Obtenha a sessão aqui
            await instance.post(
                "http://" + this.__ip + "/set_configuration.fcgi?session=" +  this._session,
                {
                monitor: {
                    request_timeout: "5000",
                    hostname: hostName,
                    port: port,
                    path: "api/notifications",
                },
                }
            );

        } catch (error) {
            console.log(error)
            throw new Error(error.message)
        }
    }

    async getSerial() {
        try {
            await this.login()
            await this.validSession()
            const response = await instance.post(
                "http://" + this.__ip + "/get_about.fcgi?session=" + this._session
            );
            return response.data;
            } catch (error) {
            throw new Error(
                "{!} Error interno ao tentar consultar a info do sistema."
            );
        }
    }
}

module.exports = Monitor;
