// control-id/objects/access_rules/AccessRules.js
const axios = require("axios");
const http = require("http");
const Login = require("../../login/Login");

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

class AccessRules {
  /**
   * @param {string} ip The IP address of the host
   * @param {string} login The login Session token 
   */
  constructor(ip, login) {
    this.__ip = ip; // HOST
    this._session = login;
  }

  /**
   * @param {number} 
   * @returns
   */
  async getAccessRules() {
    try {
      return new Promise(async (resolve, reject) => {
        try {
          const response = await instance.post(
            "http://" + this.__ip + "/load_objects.fcgi?session=" + this._session,
            {
                "object": "time_zones"
            },
            {
              headers: {
                "Content-Type": "application/json",
              },
            },
          );

          resolve(response.data);
        } catch (error) {
          reject(new Error(error.response.data.error));
        }
      });
    } catch (error) {
      throw new Error(error.message);
    }
  }

  async createAccessRules(id_user, access_rule_id) {
    try {
      return new Promise(async (resolve, reject) => {
        try {
          const response = await instance.post(
            "http://" + this.__ip + "/create_objects.fcgi?session=" + this._session,
            {
              "object": "user_access_rules",
              "values": [
                {
                  "user_id": 10050,
                  "access_rule_id": 1,
                },
              ],
            },
            {
              headers: {
                "Content-Type": "application/json",
              },
            }
          );

          resolve(response);
        } catch (error) {
          reject(new Error(error.response.data.error));
        }
      });
    } catch (error) {
      throw new Error(error.message);
    }
  }
}

// (
//   async function logagrupo(){
//     try {
  
//       const Grupo = new Rules('192.168.88.11');
  
//      console.log( await Grupo.Create_acess_rules("10050"))
      
//     } catch (error) {
//       console.error(error)
//     }
//   }
// )()
module.exports = AccessRules;
