// control-id/login/Login.js
const axios = require("axios");
const http = require("http");

const instance = axios.create({
  httpAgent: new http.Agent({ rejectUnauthorized: false }),
  headers: {
    "Content-Type": "application/json",
    flags: ["--disable-web-security", "--ignore-certificate-errors"],
  },
  insecureHTTPParser: true,
});

const MAX_ATTEMPTS = 2;

class Login {

  /** @private */ user;
  /** @private */ senha;

  constructor(ip, login, senha) {
    this.__ip = ip;
    this.user = login,
    this.senha = senha
  }

  
  async login() {
    let attempt = 0;

    while (attempt < MAX_ATTEMPTS) {
      try {
        const response = await instance.post(
          `http://${this.__ip}/login.fcgi`,
          {
            login: this.user || "admin",
            password: this.senha || "admin",
          }
        );

        this._session = response.data.session;
        return this._session;

      } catch (error) {
        attempt++;
        console.warn(`[!] Tentativa de login ${attempt} falhou: ${error.message}`);

        if (attempt >= MAX_ATTEMPTS) {
          throw new Error(`Login falhou após ${MAX_ATTEMPTS} tentativas: ${error.message}`);
        }

        // opcional: aguardar 1 segundo antes de tentar novamente
        await new Promise(res => setTimeout(res, 1000));
      }
    }
  }

  async isSessionValid(){
    try {
      await this.login();
      const response = await instance.post(
        `http://${this.__ip}/session_is_valid.fcgi?session=${this._session}`
      );
      return response.data;
    } catch (error){
      throw new Error(error.message);
    }
  }

  async changePassword(login, senha){
     try {
      await this.login();
      const response = await instance.post(
        `http://${this.__ip}/change_login.fcgi?session=${this._session}`,
        {
          login: login,
          password: senha,
        }
      );

      return response.data;
    } catch (error) {
      throw new Error(error.message);
    }
  }
}

module.exports = Login;
